Rework display to resemble JSON

kpdecker opened the pull request:
Mimic JSON structures with {key: type} and [type] structures. Extended validation data is pushed into collapsible sections to simplify the UI structure.

Additionally reduces the vertical height of the UI where possible to help with UI cohesion.

Preview of behavior with these changes (slightly older format):
![lout](https://cloud.githubusercontent.com/assets/196390/10255027/16fdaaf2-690d-11e5-8d5f-f220a0ad9492.gif)

Continuation of some of the discussion in #120

kpdecker committed:
Rework display to resemble JSON

Mimic JSON structures with {key: type} and [type] structures. Extended validation data is pushed into collapsible sections to simplify the UI structure.

Additionally reduces the vertical height of the UI where possible to help with UI cohesion.

Marsup commented:
How does it play out on highly nested scenarios ? Why only some sections unfold and others not ?

kpdecker commented:
For nested structures, the structure is always visible and the drawers open and close between the name and the children:
![image](https://cloud.githubusercontent.com/assets/196390/10263124/99ac8ec4-69a7-11e5-9551-b79e2537bbae.png)
![image](https://cloud.githubusercontent.com/assets/196390/10263125/a2e7e5ce-69a7-11e5-83c6-9383a79707ad.png)

One case where this doesn't work as well right now is try statements with alternatives with nested objects:
![image](https://cloud.githubusercontent.com/assets/196390/10263166/9ccaef78-69a8-11e5-9a49-f7462369c518.png)

I haven't really been happy with how alternatives are displayed here. I'll play around with it some more and see if I can figure out a better that exposes this data in the children section in a clear way rather than the drawer.

Marsup commented:
At what point does it become gray ? I don't understand the meaning of it.

kpdecker commented:
Gray is the drawer section and has metadata and the structural data is in
the top level white section. Goal was to make high level structure visible
immediately and then specific validation rules visible in the drill downs.
On Sat, Oct 3, 2015 at 9:14 AM Nicolas Morel <notifications@github.com>
wrote:

> At what point does it become gray ? I don't understand the meaning of it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hapijs/lout/pull/126#issuecomment-145252727>.
>

kpdecker commented:
I've updated the PR to handle alternatives in a better fashion:
<img width="285" alt="screen shot 2015-10-03 at 6 52 10 pm" src="https://cloud.githubusercontent.com/assets/196390/10265712/d7387186-6a01-11e5-86f5-2a52d37d0df6.png">

They are now treated in a similar manner to children of object values, leaving the structure displayed while pushing detailed validation information into the drawer. I used `alternatives ( )` to denote this case so it's a bit clearer that these aren't array or object structures.

Marsup commented:
Since it's a big change in terms of design, I'd like to wait for feedback from users if you don't mind. Anyone ?

dalmaer commented:
I find this much more readable than the large tree.

stefanli commented:
Love this suggestion. I find the existing format really hard to read.

Marsup commented on test/index.js:
```diff
@@ -745,7 +744,7 @@ describe('Multiple connections', function () {
     it('should load all the servers routes', function (done) {
 
-        server.inject('/docs', function (res) {
+        server.connections[0].inject('/docs', function (res) {
```
Why do you have to do that ?

kpdecker replied:
This was a breaking change in hapi. Next master build will fail as well without this change. I did this here because PRs that are red are harder to get merged :)

Marsup commented on lib/index.js:
```diff
@@ -317,15 +333,13 @@ internals.getParamsData = function (param, name) {
-                });
-            }
-        });
+        // If we have only one specific rule then set that to our type for
```
This looks a bit weird, replacing the type with a rule, in which use case does it happen ?

kpdecker replied:
This is to try to simplify the display for cases like this `Joi.string.guid()` to just display guid rather than string with a drawer for guid.

Marsup commented on lib/index.js:
```diff
@@ -352,10 +366,10 @@ internals.getExistsValues = function (type, exists) {
         }
 
-        return value;
+        return JSON.stringify(value);
```
Don't you prefer doing that in the template and keep the raw object so that people providing their own templates can use that more easily ?

kpdecker replied:
Did this here as there's no way to differentiate between the key values returned above and actual strings in the template.

Marsup replied:
Not even with a helper ?

kpdecker replied:
Not if the method above returns a string for key references and strings for string literals. The whole map function could be moved into a helper, but it doesn't seem like a huge win one way or the other and requires that people duplicate the helpers for their own templates (if my understanding of how the `helpersPath` feature works is accurate).

Marsup commented:
I've seen a few breaking changes, we'll have to bump joi peer dep. It's old enough anyway.

kpdecker commented:
What needs to happen to move forward with this?

Marsup commented:
4 tests are failing on my environment (9, 47, 48, 49), can't yet determine why.

Some routes look weird with that new format :
- /rootarray : shouldn't rules be with the array in a drawer ?
- /rootemptyobject : that's completely empty, maybe this should be removed, I don't see any real world use case for that
- /withauthasobject : alignment seems awkward
- /withpattern : fact that it's a regexp is not obvious, could as well be a string, previous version was as bad)

Also, joi just introduced ordered arrays, I'm not sure which representation is best with that style. The fact that it is already a [] kind of suggests that it's ordered but then what to do with the old unordered items ?

And we should probably bump all peer deps to their latest major.

kpdecker commented:
Unit test failures appear to be related to either the joi updates or hapi updates. After an npm update I'm seeing them as well. They are all cases where `lout` is set to false on a particular route config but we are still rendering for whatever reason.

- `/rootarray` I couldn't think of a good option for having an expander here, so I put it in an always open drawer: 
![image](https://cloud.githubusercontent.com/assets/196390/10534393/8d4d0906-739d-11e5-9fb4-9eaf52d3e425.png)

- `/rootemptyobject`: This might not make sense for the payload case, but it does for the response schema. I changed it to force the braces for empty: 
![image](https://cloud.githubusercontent.com/assets/196390/10534577/d19c07ae-739f-11e5-99d8-0a2f45bc41e5.png)

- `/withauthasobject`: Thoughts on this? 
![image](https://cloud.githubusercontent.com/assets/196390/10534215/d4b4f4d2-739a-11e5-94fc-b63040f4d886.png)
- `/withpattern`: What about this? 
![image](https://cloud.githubusercontent.com/assets/196390/10534226/14433d34-739b-11e5-9fdd-21e33008fef2.png)

- Ordered arrays: I think I like creating a flag value similar to how `/withallowunknown` works best for that case, having a `ordered` or  `unordered` flag, or potentially showing one or the other for all arrays.

Let me know what you think on these changes and I can implement/push the ones that we want to move forward with.

Marsup commented:
- /rootarray: shouldn't it be next to the 1st bracket ?
- /withauthasobject: perfect, although it seems a bit off vertically
- /withpattern perfect
- ordered is not a flag, it's a different serie of items, both can be present

I'll try to have a look at the failures, it might not be your fault.

kpdecker commented:
- rootarray: Not sure what you mean. This is the same location as it would be rendered if it were a collapsible child element (although there may be some padding differences)
- withauth: What do you think should change for vertical layout?
- order: I'm not familiar with the ordered API. What does `Joi.items(foo, bar).ordered(1, 2, 3).items(baz, bat)` validate? Is it like alternatives or are things somehow combined? Perhaps if describing the code above in human language would help us come up with a UI (and me to understand the new API :))

Marsup commented:
- rootarray : I felt it looked more natural like `[ x <- here`, not sure it is
- withauth: left and right parts don't seem exactly on the same line, right is a few px lower
- it validates an array that start with 1, 2, 3 in order, and maybe contains anything of type foo/bar/baz/bat in whatever order, is it clear like that ? Maybe that's a sign joi's docs is not good enough on this part.

kpdecker commented:
- root: Not sure it's worth additional extra code to handle that edge case, to be honest. What is there right now seems like it's good enough.
- auth: The line-heights were different causing the slightly layout difference. Fixed.
- order:
  - Since this is a new feature, I think it's better to defer that to a different pull request.
  - joi docs: It might be nice to explain how item and ordered interact with one another in API.md. Being a bit more explicit doesn't seem like a bad thing.
- Unit tests: #129 fixes the UT failures. I've rebased this PR to derive from that branch so this should pass as well.

Marsup commented:
It can indeed be a different PR but I won't release this as is if it's to change the formatting days later. Just need to find an obvious way to present that.

kpdecker commented:
Lets move the ordered array portion to #130.

Marsup commented:
I think this part is ready to merge, can you squash the commits before that ?

kpdecker commented:
Squashed and pushed
